[Function builders] Fix-Its and code completion improvements

DougGregor opened the pull request:
Extend diagnostics and code completion to help guide developers toward implementing function builders. Specifically:
* When the function builder transformation cannot be applied because of a missing `build` method, e.g., because the body uses `if-else` but the function builder doesn't provide `buildEither(first:)` and `buildEither(second:)`, add a note + Fix-It to the error message that adds stubs for the functions needed to support those statements.
* When code-completing with the definition of a function builder type, provide a full set of completions for all of the `build` functions that function builders support.
* Add an educational note outlining the builder functions that a function builder can provide.

Fixes [SR-13536](https://bugs.swift.org/browse/SR-13536) / rdar://problem/68732123

DougGregor committed:
[Function builders] Add Fix-Its for missing build* members in builders.

When a use of a function builder involves a statement kind that the
function builder doesn't support (e.g., if-else), add a note to the
diagnostic that specifies what methods need to be added to the
function builder to support that statement, including Fix-Its with
stub implementations.

DougGregor committed:
[Function builders] Add code completion for build* functions.

When performing code completion inside the declaration of a type with
the function builder attribute, also include completions for all of
the possible "build" functions.

brentdax commented on lib/Sema/CSDiagnostics.cpp:
```diff
@@ -5490,12 +5491,93 @@ SourceLoc SkipUnhandledConstructInFunctionBuilderFailure::getLoc() const {
+      componentType = inferFunctionBuilderComponentType(builder);
+    }
+
+    if (buildInsertionLoc.isInvalid()) {
```
I take it this will prevent us from showing these notes for declarations from the SDK. What about other modules in e.g. the same Xcode project? Will swiftsourceinfo files give us source locations for those?

DougGregor replied:
Yes, that's correct. `swiftsourceinfo` files should make it work with other modules in the same project/workspace.

DougGregor committed:
[Code completion] Add doc comments for function builder build functions

Add brief documentation comments to each of the code completions
produced for a build* function in a code completion.

theblixguy commented on lib/IDE/CodeCompletion.cpp:
```diff
@@ -5136,6 +5140,100 @@ class CompletionOverrideLookup : public swift::VisibleDeclConsumer {
+          "outermost block statement to produce the final returned result";
+
+    case FunctionBuilderBuildFunction::BuildLimitedAvailability:
+      return "If declaration, this will be called on the partial result of "
```
```suggestion
      return "If declared, this will be called on the partial result of "
```

DougGregor replied:
Good catch, fixed!

theblixguy commented on test/IDE/complete_function_builder.swift:
```diff
@@ -93,3 +94,21 @@ struct EnumToVoidBuilder {
+}
+
+// IN_FUNCTION_BUILDER_DECL: Begin completions, 8 items
+// IN_FUNCTION_BUILDER_DECL: Pattern/CurrNominal:                buildBlock(_ components: Any...) -> Any { {|}; name=buildBlock(_ components: Any...) -> Any {; comment=Required by every
```
Hmm, I think inserting this would conflict with existing `buildBlock` declaration. Perhaps we should only suggest adding `build*` methods that aren't explicitly declared by the user in the function builder type?

DougGregor replied:
I had thought about that, but there are a number of cases where you might want to do some overloading here, and having a redundant completion is much better than having no completion.

theblixguy replied:
Hmm, sure, that seems reasonable. What do you think about using placeholders for types if a declaration already exists? Right now,  I think inserting one would trigger a redeclaration error and using placeholders for types can help avoid that.

DougGregor replied:
I don't feel like that level of tweaking is really worthwhile. Maybe we help in a few narrow places, and maybe it's just weird because two completions in a row produce different results.

theblixguy commented on test/IDE/complete_function_builder.swift:
```diff
@@ -93,3 +94,21 @@ struct EnumToVoidBuilder {
+struct AnyBuilder {
+  static func buildBlock(_ components: Any...) -> Any { 5 }
+
+  static func #^IN_FUNCTION_BUILDER_DECL^#
```
It might be good to add another test where we don't have `static func` to test the path where we insert it if it's missing.

DougGregor replied:
Good call, and I had a bug there! Fixed.

theblixguy commented on lib/Sema/BuilderTransform.cpp:
```diff
@@ -1884,3 +1886,107 @@ bool TypeChecker::typeSupportsBuilderOp(
+  return componentType;
+}
+
+void swift::printFunctionBuilderBuildFunction(
```
It would be great to use this for the "missing buildBlock" case too in `visitFunctionBuilderAttr`!

DougGregor replied:
Absolutely! Done in https://github.com/apple/swift/pull/33972/commits/3b8656dc6fd94a91d2c1dd3e1b98343a34d0b731, along with `buildLimitedAvailability(_:)`.

rintaro commented on lib/Sema/BuilderTransform.cpp:
```diff
@@ -1884,3 +1886,107 @@ bool TypeChecker::typeSupportsBuilderOp(
+  }
+
+  if (!printedResult)
+    printer << " -> " << componentTypeString << " {";
```
Please move `" {"` to `if (stubIndent)` branch. `Builder.addBraceStmtWithCursor()` inserts full body block which is printed as ` {|}` in `swift-ide-test`. In the test cases, `-> Any { {|}` ends up with `-> Any { {\n <#code>\n }` In SourceKit. They should be just `-> Any {|}`.

DougGregor replied:
Good catch! Fixed in https://github.com/apple/swift/pull/33972/commits/23b6b409cebc43c71a0344d177f117891475166a

DougGregor committed:
[Function builders] Add an educational note on the build* functions.

Introduce an educational note with a synopsis of the build* function
declarations one can add to a function builder, and associate it with
the diagnostics indicating incorrect or missing build* functions.

DougGregor committed:
[Code completion] Fix incorrect addition of '{' in function builder results.

Noticed by Rintaro, thank you!

DougGregor committed:
[Code completion] Test and fix 'static func' for function builder methods.

DougGregor commented on include/swift/AST/DiagnosticsSema.def:
```diff
@@ -5312,6 +5312,16 @@ NOTE(function_builder_buildblock_enum_case, none,
+     "add 'buildOptional(_:)' to the function builder %0 to add support for "
+     "'if' statements without an 'else'", (Type))
+NOTE(function_builder_missing_build_either, none,
+     "add 'buildEither(first:)' and `buildEither(second:)' to the function "
```
Thanks, fixed!

DougGregor committed:
[Function builders] Add stubs for missing buildBlock and buildLimitedAvailability

The lack of these functions in a function builder is diagnosed in
different places in the type checker. Extend these diagnostics to
provide stubs in the function builder.

Thanks to Suyash for the suggestion!

DougGregor commented:
@swift-ci smoke test

DougGregor commented:
Fix-It matching with newlines doesn't work with Windows because of CRLF. Sigh.

DougGregor committed:
[Function builders] Separate out the Fix-It tests.

Testing newlines in Fix-Its does not work well with Windows, so
separate out the Fix-It tests for function builder stubs.

DougGregor commented:
@swift-ci please test

DougGregor commented:
@swift-ci please smoke test

DougGregor commented:
@swift-ci please smoke test Linux

DougGregor commented:
@swift-ci please smoke testmacOS

DougGregor commented:
@swift-ci please smoke test macOS
